### PR TITLE
Scenario template various improvements

### PR DESCRIPTION
**Pull request by NickAragua**
Implements a few improvements to scenario templates. Notably:
- "minimum unit weight" now works
- changed text on 'finalize' button to be more reflective of what it will do when the user clicks it
- transports might wind up having battle armor inside them (goes back to classic AtB too)
- clan stars generated for template scenarios might be novas with battle armor
- clans don't mix tanks and mechs until the republic of the sphere starts putting its foot down

best reviewed by individual commit

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 1]
```diff
@@ -30,6 +30,7 @@
 import java.util.UUID;
 import java.util.stream.Collectors;
 
+import org.apache.xalan.xsltc.compiler.Pattern;
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: What was this used for?

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 2]
```diff
@@ -977,48 +982,96 @@ public static Entity getInfantryEntity(UnitGeneratorParameters params, int skill
...
+                }
+                
+                if(transportedUnit == null) {
+                    transportedUnit = generateTransportedInfantryUnit(newParams, bayCapacity, skill, campaign);
                 }
 
-                MechSummary ms = campaign.getUnitGenerator().generate(newParams);
+                // sometimes something crazy will happen and we will not be able to load the unit into the transport
+                // so let's at least have it deploy at the same time as the transport
+                transportedUnit.setDeployRound(transport.getDeployRound());
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Is `transportedUnit` guaranteed to be non-null here?

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 3]
```diff
@@ -1099,6 +1108,64 @@ private static Entity generateTransportedBAUnit(UnitGeneratorParameters params,
...
+        // if yes, then pick the fastest mech and load it up, adding the generated BA to the transport relationships.
+        
+        // non-clan forces and units that aren't stars don't become novas
+        if(!Faction.getFaction(factionCode).isClan() && (starUnits.size() != 5)) {
+            return transportedUnits;
+        }
+        
+        // logic copied from AtBScenario.addStar() to randomly determine if the given unit is actually going to be a nova
+        // adjusted from 11/8 to 8/6 (distribution of novas in newest AtB doc is a lot higher) so that players actually encounter novas
+        // whatever CBS is still gets no novas, so there
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Clan Blood Spirit I'd imagine.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 4]
```diff
@@ -1099,6 +1108,64 @@ private static Entity generateTransportedBAUnit(UnitGeneratorParameters params,
...
+        } else if (factionCode.equals("CBS")) {
+            novaTarget = 13;
+        }
+        
+        if(roll < novaTarget) {
+            return transportedUnits;
+        }
+        
+        Entity actualTransport = null;
+        for(Entity transport : starUnits) {
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: We're picking the slowest one assuming its the heaviest?

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 5]
```diff
@@ -1099,6 +1108,64 @@ private static Entity generateTransportedBAUnit(UnitGeneratorParameters params,
...
+        // if we're generating a riding BA, do so now, then associate it with the designated transport
+        UnitGeneratorParameters params = new UnitGeneratorParameters();
+        params.setFaction(factionCode);
+        params.setQuality(quality);
+        params.setYear(campaign.getGameYear());
+        params.addMissionRole(MissionRole.MECHANIZED_BA);
+        params.setWeightClass(UNIT_WEIGHT_UNSPECIFIED);
+
+        Entity transportedUnit = generateTransportedBAUnit(params, IUnitGenerator.NO_WEIGHT_LIMIT, skill, campaign);
+        transportedUnit.setDeployRound(actualTransport.getDeployRound());
```
Comment (sixlettervariables) [inline MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Are we guaranteed `transportedUnit` is non-null?

**Comment by sixlettervariables** [comment]
CI failures are because your code doesn't have #1479 landed.
```
> Task :MekHQRoot:MekHQ:compileJava
/home/vsts/work/1/s/MekHQ/build/repo/mekhq/MekHQ/src/mekhq/MekHQ.java:94: error: MekHQ is not abstract and does not override abstract method gameClientFeedbackRquest(GameCFREvent) in GameListener
public class MekHQ implements GameListener {
       ^
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
1 error

```

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java (hunk 1 above)
Comment (apokrovski) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: No idea, but it was unused.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java (hunk 2 above)
Comment (apokrovski) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Probably not. I'll bullet-proof it.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java (hunk 4 above)
Comment (apokrovski) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Whoops, I wanted to put them on the fastest one.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java (hunk 5 above)
Comment (apokrovski) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Probably not.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java (hunk 3 above)
Comment (Windchild292) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Yeah, that's their tag

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java [hunk 6]
```diff
@@ -1310,6 +1463,36 @@ private static String adjustWeightsForFaction(String weights, String faction) {
...
+        // higher-rated clan units become increasingly unlikely to use vehicles
+        int vehicleTarget = 6;
+        if(factionCode.equals("CHH") || factionCode.equals("CSL") || factionCode.equals("CBS")) {
+            vehicleTarget = 8;
+        } else {
+            vehicleTarget -= forceQuality;
+        }
+        
+        // we randomly determine tank or mek
+        int roll = Compute.d6(2);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: I'm pretty sure the logic or description here is flawed. A base target of 6 on 2d6 is 27.8% (just below 2/6) whereas a base target of 4 would be 16.7% (1/6) on 2d6.

Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Unit quality theoretically ranges from 0 (F) to 5 (A*). For clans, that translates into "backwater Solamha" to "front line elite". The idea being that front line elites are too good for tanks (6 - 5 = 1 which is never on a 2d6) while dusty old warriors in the backwoods take what they can get (6 - 0 = 6+ which is sometimes). It's kind of a spitball thing. If it winds up not looking right, I'll adjust the odds going forward.

Anyway, I'll update the description.

File: MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java (hunk 4 above)
Comment (NickAragua) [reply MekHQ/src/mekhq/campaign/mission/AtBDynamicScenarioFactory.java]: Looking at it again, actual transport is the "current fastest". If actualtransport.getMP is less than the currently evaluated transport.getMP, then the currently evaluated transport becomes the actual transport by virtue of being faster.

**Comment by NickAragua** [comment]
Took me a while to figure out what the problem was, but I think the issue is that the maven repo needs to be updated with a spelling change to the gameClientFeedbackRequest method signature

**Comment by NickAragua** [comment]
Ok, I updated the maven repo, but still no compile. Giving up for now.

**Comment by Windchild292** [comment]
It doesn't appear to have run again?

**Comment by NickAragua** [comment]
Nah, I just manually re-ran it

**Comment by Windchild292** [comment]
Yeah, I checked the date and it said last run 2 days ago

**Comment by Windchild292** [comment]
Looks like LGTM ran, but the others didn't @NickAragua

**Comment by NickAragua** [comment]
Finally.

**Comment by Windchild292** [comment]
My fault on this one, sorry @NickAragua.

**Comment by NickAragua** [comment]
All's well that ends well.

**Pull request merged**